### PR TITLE
OCPERT-9 Cache slack user/group IDs in runtime

### PR DESCRIPTION
**Pull request by jhuttana**
Could you please review the changes?

### CODE REVIEW

File: oar/core/notification.py [hunk 1]
```diff
@@ -265,11 +265,15 @@ def get_user_id_by_email(self, email):
             str: slack user id
         """
         email = self.transform_email(email)
+
+        if in_memory_cache.get(email) is not None:
+            return "<@%s>" % in_memory_cache.get(email)
```
Comment (rioliu-rh) [inline oar/core/notification.py:270]: @jhuttana this line has duplicate pattern declaration, suggest to declare local var `userid`, fill the var by if/else logic
```
userid = None
if email in cache_dict:
	userid = cache_dict.get(email)
else:
	// api call
  userid = resp["user"]["id"]
  cache_dict[email] = userid

return "<@%s>" % userid
```
it looks more efficient

File: oar/core/notification.py [hunk 2]
```diff
@@ -287,6 +291,8 @@ def get_group_id_by_name(self, name):
             group id: slack group id
         """
         ret_id = ""
+        if in_memory_cache.get(name) is not None:
```
Comment (rioliu-rh) [inline oar/core/notification.py:294]: @jhuttana same concern as above

File: oar/core/notification.py (hunk 1 above)
Comment (rioliu-rh) [reply oar/core/notification.py:270]: also, we can add debug level log here to indicate the contact is retrieved from cache. so if you execute the cmd with option `-v/--debug`, the logic can be traced.

**Comment by jhuttana** [comment]
Thanks for the comments Rio. I will address these comments.

File: oar/core/notification.py [hunk 3]
```diff
@@ -238,6 +237,8 @@ def send_email(self, to_addrs, subject, content):
 
 
 class SlackClient:
+    cache_dict = dict()
```
Comment (rioliu-rh) [inline oar/core/notification.py:240]: @jhuttana cache object should be moved to class instructor `__init__` like `self.cache = {}`, all subsequent calls need to be changed as well. e.g. `self.cache[xx]=y`

Comment (jhuttana) [reply oar/core/notification.py:240]: That would be perfect. I will do that.

File: oar/core/notification.py (hunk 1 above)
Comment (jhuttana) [reply oar/core/notification.py]: Sure

File: oar/core/notification.py [hunk 4]
```diff
@@ -239,6 +238,7 @@ def send_email(self, to_addrs, subject, content):
 
 class SlackClient:
     def __init__(self, bot_token):
+        self.cache_dict = dict()
```
Comment (rioliu-rh) [inline oar/core/notification.py:241]: can we do code formatting, put all the local vars together
```
self.a = xx
self.b = yy
```

File: oar/core/notification.py [hunk 5]
```diff
@@ -287,23 +293,28 @@ def get_group_id_by_name(self, name):
...
+                            break
+            except SlackApiError as e:
+                raise NotificationException(
+                    f"query group id by name {name} error") from e
 
         if not ret_id:
             raise NotificationException(
                 f"cannot find slack group id by name {name}")
 
+        logger.debug(f"Cached group ids: {self.cache_dict}")
```
Comment (rioliu-rh) [inline oar/core/notification.py:317]: @jhuttana same concern as above https://github.com/openshift/release-tests/pull/357/files#r1848102656

File: oar/core/notification.py [hunk 6]
```diff
@@ -265,15 +265,21 @@ def get_user_id_by_email(self, email):
...
+                resp = self.client.api_call(
+                    api_method="users.lookupByEmail", params={"email": email}
+                )
+                userid = resp["user"]["id"]
+                self.cache_dict[email] = userid
+            except SlackApiError as e:
+                logger.warning(f"cannot get slack user id for <{email}>: {e}")
+                return email
 
+        logger.debug(f"Cached email ids: {self.cache_dict}")
```
[the remark this answers is not in the document]
Comment (jhuttana) [reply oar/core/notification.py:282]: I think I am not getting it.
Do you mean to change it as: `logger.debug(f"Cached email ids: {self.cache_dict.get(email)}")`  ?

[the remark this answers is not in the document]
Comment (rioliu-rh) [reply oar/core/notification.py:282]:
```
if email in cache:
	// get userid from cache
  logger.debug(f"slack user id of {email} is retrieved from cached")
else:
  // get userid from api call
  // put the userid to cache object
  logger.debug(f"slack user id of {email} is cached")
```

File: oar/core/notification.py (hunk 4 above)
Comment (jhuttana) [reply oar/core/notification.py:241]: For me it looks like declaring and initializing the runtime cache variable inside the constructor is fine (though it is referenced in two different methods).  I don't see any difference in moving its declaration to the respective methods. For eg:
```
#Case-1: Declare and initialize the variable in constructor
class test:
    def __init__(self):
        self.attr1 = dict()
    
    def a(self):
        self.attr1 = {1:'one'}
        print(self.attr1) 
    
    def b(self):
        self.attr1 = {2:'two'}
        print(self.attr1)

#Case-2: Declare and initialize the variable in respective methods
class test1:
    def __init__(self):
        print("Inside test1 class")
    
    def aa(self):
        self.attr1 = dict()
        self.attr1 = {1:'one'}
        print(self.attr1) 
    
    def bb(self):
        self.attr1 = dict()
        self.attr1 = {2:'Two'}
        print(self.attr1) 
 
#For Case-1      
t = test()
t.a()  
t.b()

#For Case-2
tt = test1()
tt.aa()
tt.bb()
```
From your comment my understanding is that you are suggesting to go for  Case-2 in the above example.
Though Case-1 and Case-2 are trying to achieve the same thing ,  Case-1 looks more efficient to me.
If my understanding is wrong could you please clarify what exactly are you suggesting?

Comment (rioliu-rh) [reply oar/core/notification.py:241]: I just want to make the code looks clean, put the all the local var declaration code together. 
```suggestion
				if not bot_token:
            raise NotificationException("slack bot token is not available")
        self.client = WebClient(token=bot_token)
        self.cache_dict = dict()
        
```

**Comment by rioliu-rh** [comment]
I think you need to create a new unit test to run funcs `get_group_id_by_name`/`get_user_id_by_email` multiple times to verify.
sample test case: https://github.com/openshift/release-tests/blob/35262b0be5f4c065b8356a4f3930864596841027/tests/test_notification.py#L24
enabled debug logging: https://github.com/openshift/release-tests/blob/35262b0be5f4c065b8356a4f3930864596841027/tests/test_artifacts.py#L9-L13

**Comment by jhuttana** [comment]
> I think you need to create a new unit test to run funcs `get_group_id_by_name`/`get_user_id_by_email` multiple times to verify. sample test case:
> 
> https://github.com/openshift/release-tests/blob/35262b0be5f4c065b8356a4f3930864596841027/tests/test_notification.py#L24
> 
> enabled debug logging:
> 
> https://github.com/openshift/release-tests/blob/35262b0be5f4c065b8356a4f3930864596841027/tests/test_artifacts.py#L9-L13

Sure I will give it a try

**Comment by rioliu-rh** [comment]
you need to enable debug level logging in module `test_notification.py`, so you can see whether the user/group id can be retrieved from cache

**Comment by jhuttana** [comment]
> you need to enable debug level logging in module `test_notification.py`, so you can see whether the user/group id can be retrieved from cache

Sure will do that

**Comment by rioliu-rh** [comment]
> > you need to enable debug level logging in module `test_notification.py`, so you can see whether the user/group id can be retrieved from cache
> 
> Sure will do that

if unit test is pass, we can merge this PR

**Comment by jhuttana** [comment]
Verify the unittest case
```
$ python -m unittest test_notification.TestNotificationManager.test_get_slack_user_id_from_cache
.
----------------------------------------------------------------------
Ran 1 test in 5.943s

OK

$ python -m unittest test_notification.TestNotificationManager.test_get_slack_group_id_from_cache
.
----------------------------------------------------------------------
Ran 1 test in 9.199s

OK

```

**Pull request merged**